crates/subspace-service/src/sync_from_dsn/fast_sync.rs

```diff
@@ -332,37 +354,43 @@ where
                     state: None,
                 })
             })
-            .collect::<Result<Vec<_>, Error>>()?;
+            .collect::<Result<Vec<IncomingBlock<Block>>, Error>>()?;
 
-        let block_to_import = blocks_to_import
-            .pop()
-            .expect("At least one block is present; qed");
+        let maybe_last_block_to_import = blocks_to_import.pop();
 
         if !blocks_to_import.is_empty() {
+            let last_queued_block_number = *blocks_to_import
+                .last()
+                .expect("Not empty; qed")
+                .header
+                .as_ref()
+                .expect("Always set; qed")
+                .number();
+
             self.import_queue_service
                 .lock()
                 .await
                 .import_blocks(BlockOrigin::NetworkInitialSync, blocks_to_import);
-        }
 
-        // Block import delay
-        // We wait to import for all the blocks from the segment except the last one
-        // TODO: Replace this hack with actual watching of block import
-        self.wait_for_block_import(second_last_block_number.into())
-            .await;
+            // Block import delay
+            // We wait to import for all the blocks from the segment except the last one
+            // TODO: Replace this hack with actual watching of block import
+            self.wait_for_block_import(last_queued_block_number).await;
+        }
 
         // Notify Substrate sync by importing block with using BlockOrigin::NetworkBroadcast
+        if let Some(last_block_to_import) = maybe_last_block_to_import {
```

Comment:
This indentation is unnecessary because the last block must always be present.

Reply:
As explained above it is not actualyl the case. We might have the first and only block inserted raw and nothing after that. I tried to explain this in PR description.